src/control/lib/control/auto.go

```diff
@@ -449,3 +475,55 @@ func (req *ConfigGenerateReq) checkStorage(ctx context.Context, resp *ConfigGene
 
 	return nil
 }
+
+// checkCpus validates VOS target count and xstream helper thread count
+// recommended values and populates ioserver storage config.
+//
+// The target count should be a multiplier of the number of SSDs and typically
+// daos gets the best performance with 16x targets per I/O Server so target
+// count will be between 12 and 16. This allows to cover 1 to 8 SSDs which is
+// the supported range.
+//
+// Validate number of targets + 1 cores are available per IO servers, not
+// usually a problem as sockets normally have at least 18 cores.
+//
+// Create helper threads for the remaining available cores, e.g. with 24 cores,
+// allocate 7 helper threads. Number of helper threads should never be more than
+// number of targets.
+func (req *ConfigGenerateReq) checkCPUs(ctx context.Context, coresPerNuma int, resp *ConfigGenerateResp) error {
+	for _, ioCfg := range resp.ConfigOut.Servers {
+		var numTargets int
+		switch ioCfg.Storage.Bdev.DeviceCount {
+		case 1, 2, 4, 8:
+			numTargets = 16
+		case 5:
+			numTargets = 15
+		case 7:
+			numTargets = 14
+		case 3, 6:
+			numTargets = 12
+		}
```

Comment:
Seems like there should be a default case here. What happens if there are more devices than 8?

Reply:
I removed the specific validation from this function because it already happens in validateNvmeStorage() https://github.com/daos-stack/daos/pull/3889/files/1677c4c990ef584e579e38230c23fb24de31ecf6#diff-a723789e4c16a154f22560c66bff2424c7b75c78231333a593aa878cbcf330d8R377

Reply:
If it can't happen, maybe the default should return an error. I was going to suggest a panic, but that might get us in more trouble if we make a coding error in the future and tests don't catch it.

Reply:
Would be good to have some clarification on this as we have some inconsistencies in the code base regarding the use of panic(), I noticed one was just introduced in the multi-replica PR. Sometimes when a situation "really shouldn't happen" we return an error and occasionally we panic(). Obviously we don't want to panic too much but should we be panicking at all?

Reply:
yes I agree with that and it's the approach I've been following